gateway-service/src/main/java/org/zowe/apiml/gateway/security/config/JWTLogoutHandler.java

```diff
@@ -34,21 +36,27 @@
     public void logout(HttpServletRequest request, HttpServletResponse response, Authentication authentication) {
         Optional<String> token = authenticationService.getJwtTokenFromRequest(request);
         try {
-            checkJwtTokenFormat(failure, request, response, token);
+            if (token.isPresent()) {
+                invalidateJwtToken(failure, request, response, token.get());
+            } else {
+                failure.onAuthenticationFailure(request, response, new TokenNotProvidedException("The token you are trying to logout is not present in the header"));
+            }
         } catch (ServletException e) {
             log.error("The response cannot be written during the logout exception handler: {}", e.getMessage());
         }
     }
 
-    private void checkJwtTokenFormat(FailedAuthenticationHandler failure, HttpServletRequest request, HttpServletResponse response, Optional<String> token) throws ServletException {
-        if (token.isPresent()) {
+    private void invalidateJwtToken(FailedAuthenticationHandler failure, HttpServletRequest request, HttpServletResponse response, String token) throws ServletException {
+        if (authenticationService.isInvalidated(token)) {
+            failure.onAuthenticationFailure(request, response, new TokenNotValidException("The token you are trying to logout is not valid"));
+        } else {
             try {
-                authenticationService.invalidateJwtToken(token.get(), true);
-            } catch (TokenNotValidException e) {
-                failure.onAuthenticationFailure(request, response, new TokenFormatNotValidException(e.getMessage()));
+                authenticationService.invalidateJwtToken(token, true);
+            } catch (TokenFormatNotValidException e) {
+                failure.onAuthenticationFailure(request, response, e);
+            } catch (Exception e) {
+                failure.onAuthenticationFailure(request, response, new AuthenticationTokenException("Error while logging out token"));
```

Comment:
I think this is causing some problem on the error handling. I've tried to intentionally pass an invalid token via Cookie (I deleted part of the token) and call the `/logout` endpoint. On master it returns `400` with `TokenFormatNotValidException` exception, which is the expected behaviour. Now it returns `500` with:
```
{
    "messages": [
        {
            "messageType": "ERROR",
            "messageNumber": "ZWEAG100E",
            "messageContent": "Authentication exception: 'Error while logging out token' for URL '/api/v1/gateway/auth/logout'",
            "messageKey": "org.zowe.apiml.security.generic"
        }
    ]
}
```
It is because `authenticationService.invalidateJwtToken(token, true);` is throwing `TokenNotValidException` and you are catching a generic `Exception e` which will be throw this message, which is not correct in this case. It should return instead 400 with `TokenFormatNotValidException`, that is how it was before. 
Therefore I think you can add a catch:
```
 } catch (TokenNotValidException e) {
                failure.onAuthenticationFailure(request, response, new TokenFormatNotValidException(e.getMessage()));
            }
```
and keep the last generic catch to catch other possible exception such as `ServiceNotAccessibleException`

Reply:
Added `TokenNotValidException` catch that returns `TokenFormatNotValidException`, and included `AuthenticationException` catch as well as generic `Exception` catch to provide error handling for a specific instance of `AuthenticationException` as well as handle any other errors that happen (like `ServiceNotAccessibleException`).